app/controllers/visualizations_controller.rb

```diff
@@ -126,9 +126,67 @@ def shorten_url
     }, status: :internal_server_error
   end
 
+  # START OF HEATMAP METHODS
+
+  def heatmap
```

Comment:
I moved all action methods from the `HeatmapHelper` to the controller itself so that the helper could be a pure static class.